x/shield/keeper/pool.go

```diff
@@ -41,122 +123,70 @@ func (k Keeper) CreatePool(ctx sdk.Context, creator sdk.AccAddress,
-			"available %s, shield %s", provider.Available, shield)
+	totalCollateral := k.GetTotalCollateral(ctx)
+	totalWithdrawing := k.GetTotalWithdrawing(ctx)
+	totalShield := k.GetTotalShield(ctx)
+	if totalShield.Add(shieldAmt).GT(totalCollateral.Sub(totalWithdrawing)) {
+		return types.Pool{}, types.ErrNotEnoughCollateral
 	}
-	provider.Available = provider.Available.Sub(shieldAmt)
-
-	endTime := ctx.BlockTime().Add(poolLifeTime)
-	id := k.GetNextPoolID(ctx)
-	depositDec := types.MixedDecCoinsFromMixedCoins(deposit)
 
-	pool := types.NewPool(shield, shieldAmt, depositDec, sponsor, sponsorAddr, endTime, id)
+	// Check pool shield limit.
+	poolParams := k.GetPoolParams(ctx)
+	maxShield := totalCollateral.Sub(totalWithdrawing).ToDec().Mul(poolParams.PoolShieldLimit).TruncateInt()
+	if shieldAmt.GT(maxShield) {
+		return types.Pool{}, types.ErrPoolShieldExceedsLimit
+	}
 
-	// Transfer deposit to the Shield module account.
-	if err := k.DepositNativePremium(ctx, deposit.Native, creator); err != nil {
+	// Transfer service fees to the Shield module account.
+	if err := k.DepositNativeServiceFees(ctx, serviceFees.Native, creator); err != nil {
 		return types.Pool{}, err
 	}
 
-	// Make a pseudo-purchase for B.
-	purchaseID := k.GetNextPurchaseID(ctx)
-	expirationTime := pool.EndTime.Add(-k.gk.GetVotingParams(ctx).VotingPeriod * 2)
-	purchase := types.NewPurchase(purchaseID, shield, ctx.BlockHeight(), expirationTime, expirationTime, expirationTime, "shield for sponsor")
-
+	// Set the new project pool.
+	id := k.GetNextPoolID(ctx)
+	pool := types.NewPool(id, description, sponsor, sponsorAddr, shieldAmt)
 	k.SetPool(ctx, pool)
 	k.SetNextPoolID(ctx, id+1)
-	k.SetProvider(ctx, admin, provider)
-	k.SetCollateral(ctx, pool, admin, types.NewCollateral(pool, admin, shieldAmt))
 
+	// Update service fees in the global pool.
+	serviceFeesUpdate := k.GetServiceFees(ctx)
+	serviceFeesUpdate = serviceFeesUpdate.Add(types.MixedDecCoinsFromMixedCoins(serviceFees))
+	k.SetServiceFees(ctx, serviceFeesUpdate)
+
+	// Make a pseudo-purchase for B.
+	purchaseID := k.GetNextPurchaseID(ctx)
+	protectionEndTime := ctx.BlockTime().Add(protectionPeriod)
+	purchase := types.NewPurchase(purchaseID, protectionEndTime, "shield for sponsor", shieldAmt)
+	k.InsertPurchaseQueue(ctx, types.NewPurchaseList(id, sponsorAddr, []types.Purchase{purchase}), protectionEndTime.Add(k.GetPurchaseDeletionPeriod(ctx)))
 	k.AddPurchase(ctx, id, sponsorAddr, purchase)
-	k.InsertPurchaseQueue(ctx, types.NewPurchaseList(id, sponsorAddr, []types.Purchase{purchase}), expirationTime)
 	k.SetNextPurchaseID(ctx, purchaseID+1)
 
 	return pool, nil
 }
 
-// UpdatePool updates pool info and sponsor's shield.
-func (k Keeper) UpdatePool(ctx sdk.Context, updater sdk.AccAddress, shield sdk.Coins, deposit types.MixedCoins, id uint64, addTime time.Duration, description string) (types.Pool, error) {
+// FIXME UpdatePool only updates descriptions now. Any other things to be updated?
```

Comment:
Update pool is almost exactly same as createpool now. Refactor and share most logic